mon: Trim config-history to avoid MON_DISK_BIG

yzhan298 opened the pull request:
Remove the old config history entries that can accumulate in the mon.store which causes the MON_DISK_BIG issue.

Fixes: https://tracker.ceph.com/issues/67453





<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)
</details>

dvanders reviewed (changes requested):
I think this should be implemented in ConfigMonitor.cc? There is logic there about config changesets and the `config log` command is also related.
If possible, find a way to remove up to ~100 or so kv entries per trim(). More than that might be disruptive to the mon.

dvanders commented on src/mon/KVMonitor.cc:
```diff
@@ -109,6 +117,73 @@ void KVMonitor::get_store_prefixes(set<string>& s) const
+  map<int, int> version_key_deletion_count;
+
+  for (auto& [config_key, version_map] : config_versions) {
+    if (version_map.size() <= MAX_HISTORY)
```
looks good!

JoshuaGabriel commented:
jenkins test make check

JoshuaGabriel commented:
jenkins test make check arm64

ifed01 commented:
jenkins test make check

anthonyeleven commented on src/common/options/global.yaml.in:
```diff
@@ -1880,6 +1880,15 @@ options:
+- name: mon_config_history_size
+  type: size
+  level: advanced
+  desc: Defines the number of config history entries to be kept in monitor store 
```
Suggest

desc: Defines the number of config history entries to be kept in the Monitors' store

anthonyeleven commented on src/common/options/global.yaml.in:
```diff
@@ -1880,6 +1880,15 @@ options:
+  type: size
+  level: advanced
+  desc: Defines the number of config history entries to be kept in monitor store 
+  fmt_desc: The maximum number of config history
```
Truncated?

ifed01 commented:
jenkins test make check arm64

ifed01 commented:
jenkins test make check

ifed01 commented:
jenkins test make check

anthonyeleven commented on src/common/options/global.yaml.in:
```diff
@@ -1880,6 +1880,15 @@ options:
+- name: mon_config_history_size
+  type: size
+  level: advanced
+  desc: Defines the number of config history entries to be kept in the Monitors' store 
```
Thank you for the plural posessive!
And thanks to @dvanders for having alerted me to the fact that we haven't been trimming these.  I have a customer right now who I think is suffering from this very issue, most likely led by the preponderance of `osd_memory_target` entries

If I understand the code correctly, I think we would do well to clarify a bit, something like:

>desc: The number of config history changes to be retained for each central configuration option.  This includes manual changes via 'ceph config' commands and adjustments to 'osd_memory_target' by the autotuner.

anthonyeleven commented on src/mon/ConfigMonitor.cc:
```diff
@@ -980,3 +986,77 @@ void ConfigMonitor::check_all_subs()
+  map<int, int> version_key_deletion_count;
+
+  const size_t mon_config_history_size = g_conf()->mon_config_history_size;
+  const size_t max_entries_to_remove = 100; // num of records to be removed in a single tick
```
Nice.  One of my first thoughts was as to the impact of trimming tens of thousands of osd_memory_target entries the first time.

anthonyeleven commented on src/common/options/global.yaml.in:
```diff
@@ -1880,6 +1880,15 @@ options:
   - mon
   with_legacy: true
+- name: mon_config_history_size
+  type: size
```
Should this maybe be `uinit` or `int`? What happens if `0` or `-1` are entered?  Since various other options have special behaviors for those values, users might enter them out of habit.  Do we *want* an option to configure the current, unlimited behavior?    I'm half tempted to suggest a separate setting for osd_memory_target given the autotuner, but I really hate hardcoding like that.

markhpc reviewed (commented):
High level I think it makes sense.  The regex might be a bit opaque without comments.  I'd like Joao's take on it since he's more familiar with the Mon code.

Sean10 commented on src/mon/ConfigMonitor.cc:
```diff
@@ -980,3 +986,81 @@ void ConfigMonitor::check_all_subs()
+  }
+
+  for (const auto& key : keys_to_delete) {
+    mon.kvmon()->enqueue_rm(key);
```
Is there any like delete_range op?

yzhan298 commented:
jenkins test make check

yzhan298 commented:
This PR has been stayed open for a long time. Could someone from the admin group take a look? All requested changes were committed.

tchaikov commented on src/mon/ConfigMonitor.cc:
```diff
@@ -980,3 +986,81 @@ void ConfigMonitor::check_all_subs()
+  for (const auto& key : keys_to_delete) {
+    mon.kvmon()->enqueue_rm(key);
+  }
+  mon.kvmon()->propose_pending();
```
in this function's caller -- `ConfigMonitor::tick()`, if `changed` is `false`, kvmon' `have_pending` is still true, and `propose_pending()` gets called, and encodes an empty delta into paxos, even if `keys_to_delete` is empty. this is a waste to perform an unnecessary paxos round every single tick.

yzhan298 replied:
Good catch.

tchaikov commented on src/mon/ConfigMonitor.cc:
```diff
@@ -980,3 +986,81 @@ void ConfigMonitor::check_all_subs()
+    size_t remove_count = version_map.size() - mon_config_history_size;
+    auto it = version_map.begin();
+
+    for (size_t i = 0; i < remove_count && it != version_map.end() && !should_stop(); ++i, ++it) {
```
better off using range-based for loop.

yzhan298 replied:
Makes sense.

tchaikov reviewed (changes requested):
left a few comments.

yzhan298 commented on src/mon/ConfigMonitor.cc:
```diff
@@ -980,3 +986,81 @@ void ConfigMonitor::check_all_subs()
+
+    for (size_t i = 0; i < remove_count && it != version_map.end() && !should_stop(); ++i, ++it) {
+      int version = it->first;
+      for (const auto& key : it->second) {
```
done.

yzhan298 commented on src/mon/ConfigMonitor.cc:
```diff
@@ -980,3 +986,81 @@ void ConfigMonitor::check_all_subs()
+      int version = it->first;
+      for (const auto& key : it->second) {
+        mark_for_deletion(key, version);
+        if (should_stop()) break;
```
Added

yzhan298 commented on src/mon/ConfigMonitor.cc:
```diff
@@ -980,3 +986,81 @@ void ConfigMonitor::check_all_subs()
+  map<int, int> version_key_deletion_count;
+
+  const size_t mon_config_history_size = g_conf()->mon_config_history_size;
+  const size_t max_entries_to_remove = 100; // num of records to be removed in a single tick
```
Yeah, makes sense, but constexpr would be better for this case.

yzhan298 commented on src/mon/ConfigMonitor.cc:
```diff
@@ -980,3 +986,81 @@ void ConfigMonitor::check_all_subs()
+  const string history_prefix = "config-history/";
+  // key_re is used to match key entries in config-history list
+  // example: config-history/100/+osd/host:dx-arsenalceph01rack02data-01/osd_memory_target"
+  const regex key_re(history_prefix + R"((\d+)/[+-]([^/]+/[^/]+))");
```
Makes sense.

yzhan298 commented on src/mon/ConfigMonitor.cc:
```diff
@@ -980,3 +986,81 @@ void ConfigMonitor::check_all_subs()
+  const string history_prefix = "config-history/";
+  // key_re is used to match key entries in config-history list
+  // example: config-history/100/+osd/host:dx-arsenalceph01rack02data-01/osd_memory_target"
+  const regex key_re(history_prefix + R"((\d+)/[+-]([^/]+/[^/]+))");
```
Good advice, I'll make the change.

nhoad commented on src/mon/ConfigMonitor.cc:
```diff
@@ -756,12 +760,17 @@ void ConfigMonitor::tick()
-    mon.kvmon()->propose_pending();
-    paxos.unplug();
-    propose_pending();
+  if (mon.kvmon()->is_writeable()) {
+    if (_trim_config_history()) {
+      changed = true;
+    }
+    if (changed) {
+      paxos.plug();
+      encode_pending_to_kvmon();
+      mon.kvmon()->propose_pending();
+      paxos.unplug();
+      propose_pending();
+    } 
   }
```
What are the implications of this taking a long time (e.g. the first time it runs)?

yzhan298 replied:
While the deletions are capped at 100 per tick to protect the Paxos payload, the initial iteration and regex matching phase is unbounded. If a cluster has a long config history backlog, the initial run can take long time to finish. Maybe I should have a budge for reading too.

yzhan298 replied:
I thought about this problem for a while, and I still prefer to not have a complex bookmark mechanism to limit the scanning.If it takes long time(I don't think this will take more than few seconds) for the first time it runs, it's  acceptable. Another way to bypass the problem is removing the old records manually before you turn on the auto trimming service.

anthonyeleven commented:
Aw man I really want to see this.  

If the author isn’t around to carry it over the line, any chance @tchaikov could take it over?

xiubli commented:
jenkins test make check arm64

xiubli commented:
jenkins test make check

xiubli commented:
jenkins test docs

yzhan298 committed:
mon: trim config-history to avoid MON_DISK_BIG

Each config change writes a config-history/<version>/ entry into the
mon store, and nothing ever removed them. On clusters where an option
is adjusted repeatedly -- an autotuner driving osd_memory_target being
the common case -- these entries grow without bound until the monitors
report MON_DISK_BIG.

Trim the history from ConfigMonitor::tick(), retaining the most recent
mon_config_history_size versions of each option and dropping the older
ones.

Fixes: https://tracker.ceph.com/issues/67453
Co-authored-by: Kefu Chai <tchaikov@gmail.com>
Signed-off-by: Steven Zhang <yzhan298@gmail.com>

dvanders commented:
Please add a test for this.

yzhan298 commented:
> Please add a test for this.

Added.

tchaikov commented on src/mon/ConfigMonitor.cc:
```diff
@@ -761,12 +765,17 @@ void ConfigMonitor::tick()
+      mon.kvmon()->propose_pending();
+      paxos.unplug();
+      propose_pending();
+    } 
```
so, `_trim_config_history()` returning true commits the deletions through the config-change path, which always stamps a new history marker. Every trim proposal therefore leaves an empty `config-history/<N>/` that nothing can reclaim.

a narrow fix is to skip the marker when a transaction yields no history entries, and better off keeping deletion off the config-change path, like
```c++
if (changed) {
    paxos.plug();
    encode_pending_to_kvmon();
    mon.kvmon()->propose_pending();      // carries the trims along
    paxos.unplug();
    propose_pending();
} else if (_trim_config_history()) {
    // enqueues rm into kvmon only
    mon.kvmon()->propose_pending();      // deletions only
}
```

but the way how this pull request trims config history could 

- break `config reset` quietly, as only a subset of changes are reverted
- marker might not match with the removed entries, and leaves orphans
- scans the *whole* history each tick

so, i think a better way is to trim the history config entries with the granularity of changeset. to be more specific, we can make `mon_config_history_size` a count of retained changsets rather than per key, and keep a low water mark.

- store `config-history-first`, trim while `version - first > N`
- each round delete at most K whole versions from `first` up: `lower_bound("config-history/<v>/")`, remove the marker and every +/- under that exact prefix, advance `first`. This is the scan `load_changeset()` already does, so the work is O(deleted) rather than `O(total)`, and the persisted mark makes it resumable across ticks, which answers @dvanders ' question at the top of this thread
- clamp `config log` at `first`, and have `config reset` refuse below it with a message saying the history was trimmed, so we are explicit on something we cannot do.
